### PR TITLE
Ryan misc changes

### PR DESCRIPTION
**Pull request by RyanMah123**
Does the following:
* Adjusts thread periodicities to achieve worst case 7ms response time with <70% CPU utilization
* Moves CAN heartbeats to a different thread so it can have a slower periodicity than the state machine
* Implements toggling of the BIST using the on-board button
* Switches Coulomb counting to use the HAL sys tick to keep track of elasped milli-seconds

### CODE REVIEW

**Comment by RyanMah123** [comment]
Note: PLEASE REVIEW CAREFULLY, I'm not too confident on this one.

File: STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp:20]: Should you be initializing all of the class members (The R,G,B, rgb blink, rgb on, cnt) to avoid unexpected behaviour?

File: STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp:51]: Maybe do something like `DEBUG_GPIO_Port->ODR ^= DEBUG_Pin * (~cnt); ` and then after do `LEDThread::cnt = (LEDThread::cnt + 1) % 4;` and change it to a uint8? this also just avoids any far off possibility of an overflow?

File: STM32F405RGTx/Core/Src/rtos_threads/bist_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/bist_thread.cpp]: Is it possible to get the Pin name to be `BIST_Pin` or something like that?

File: STM32F405RGTx/Core/Inc/LED_thread.hpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Inc/LED_thread.hpp:5]: Is there a reason we're not doing the LEDStatus here like in MC? I think it's best to keep it consistent across the boards. For something as common as and LED Thread, maybe we can somehow implement this in the util library and you would just have to add an LED pin and port into the initialize function

File: STM32F405RGTx/Core/Src/rtos_threads/coulomb_counting_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/coulomb_counting_thread.cpp:39]: Can you just move this to line 26 and instead of HAL_GetTick, use curr_time to replace prev time

File: STM32F405RGTx/Core/Src/rtos_threads/coulomb_counting_thread.cpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Src/rtos_threads/coulomb_counting_thread.cpp:22]: Maybe rename this from get charge to get used charge or something cause that's really what its doing

File: STM32F405RGTx/Core/Inc/threads.hpp
Comment (GuyBlumenthal) [inline STM32F405RGTx/Core/Inc/threads.hpp:12]: Personally, I'm not a fan of a file not human readable on github. This is fine but do you think we could instead do this with python and out put a txt, md or json file from the script?. Excel is fine though if you think its just a lot simpler

**Review by GuyBlumenthal — COMMENTED** [review]
Just a few comments. I think it looks good overall though

File: STM32F405RGTx/Core/Inc/LED_thread.hpp
Comment (RyanMah123) [reply STM32F405RGTx/Core/Inc/LED_thread.hpp:5]: I would be down to put it in util, but I was honestly just lazy tbh lol.

File: STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp
Comment (RyanMah123) [reply STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp:20]: ye good catch

File: STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp
Comment (RyanMah123) [reply STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp:51]: When an unsigned int overflows it'll just reset back to 0, so I don't think this is an issue.

Comment (RyanMah123) [reply STM32F405RGTx/Core/Src/rtos_threads/LED_thread.cpp:51]: image

File: STM32F405RGTx/Core/Src/rtos_threads/coulomb_counting_thread.cpp
Comment (RyanMah123) [reply STM32F405RGTx/Core/Src/rtos_threads/coulomb_counting_thread.cpp:39]: I'm not sure what you mean here sry. Also, what would be the advantage?

File: STM32F405RGTx/Core/Inc/threads.hpp
Comment (ren-andy) [reply STM32F405RGTx/Core/Inc/threads.hpp:12]: or maybe leave it as a csv? so that it's viewable in excel too

**Review by ren-andy — DISMISSED** [review]
lgtm

**Comment by ren-andy** [comment]
not sure if I should be the one approving this, but lgtm

File: STM32F405RGTx/Core/Inc/threads.hpp
Comment (RyanMah123) [reply STM32F405RGTx/Core/Inc/threads.hpp:12]: There are formulas tho, CSVs can't have those

**Review by Nushaab-Syed — APPROVED** [review]
Reviewed and looks good, and completes some of the outstanding BMS G5.5 tasks we had:
- [Schedulability analysis + improvements](https://github.com/waterloop/BMS_SW_G5/pull/25/commits/8bdcd411871c3633b5eeca27c3cf53de43dc263e)
- [Timer Improvements; coulomb counting optimization](https://github.com/waterloop/BMS_SW_G5/pull/25/commits/8bdcd411871c3633b5eeca27c3cf53de43dc263e)

These will be updated on our burndown list sheet once the PR is merged. There were a few comments, like [this](https://github.com/waterloop/BMS_SW_G5/pull/25/files#r857173127) that weren't responded to which you might want to consider, thought they aren't relevant to functionality.

File: STM32F405RGTx/Core/Inc/LED_thread.hpp
Comment (GuyBlumenthal) [reply STM32F405RGTx/Core/Inc/LED_thread.hpp:5]: @Nushaab-Syed @ren-andy Can you add this as an item somewhere for an improvement in the future? An LED thread could definitely be put in the util library. Maybe you can think of other stuff we could add to the util library. Besides that, this convo is resolved.

File: STM32F405RGTx/Core/Src/rtos_threads/coulomb_counting_thread.cpp
Comment (GuyBlumenthal) [reply STM32F405RGTx/Core/Src/rtos_threads/coulomb_counting_thread.cpp:39]: oh nevermind im stupid this is outside the while loop

**Review by GuyBlumenthal — APPROVED** [review]
Ok, looks good to me. There's just one comment that stilll applies, I think the pin should be renamed so if you could do that, but I don't think that should block the merge.

File: STM32F405RGTx/Core/Inc/LED_thread.hpp
Comment (Nushaab-Syed) [reply STM32F405RGTx/Core/Inc/LED_thread.hpp:5]: Would this be suitable as a G5.5 or G6 improvement? I'll add it to the planning package accordingly. @GuyBlumenthal

**Comment by RyanJMah** [comment]
> Ok, looks good to me. There's just one comment that stilll applies, I think the pin should be renamed so if you could do that, but I don't think that should block the merge.

The name is fine to me, since it tells you that its configured as an `EXT` interrupt for a button. I'll add a `#define BIST_Pin BUTTON_EXT11_Pin` tho.

After that, I'll merge.

**Pull request merged**
